source/server/connection_handler_impl.cc

```diff
@@ -95,6 +95,35 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
           details->typed_listener_)) {
     tcp_listener_map_by_address_.insert_or_assign(
         config.listenSocketFactory().localAddress()->asStringView(), details);
+
+    auto& address = details->address_;
+    // If the address is IPv6 and isn't v6only, parse out the ipv4 compatible address from the ipv6
+    // address and put an item to the map. Then this allows the `getBalancedHandlerByAddress`
+    // can match the ipv4 any-address also.
+    if (address->type() == Network::Address::Type::Ip &&
+        address->ip()->version() == Network::Address::IpVersion::v6 &&
+        !address->ip()->ipv6()->v6only()) {
+      if (address->ip()->isAnyAddress()) {
+        // Since both "::" with ipv4_compat and "0.0.0.0" can be supported.
+        // So use `emplace` here, if there already one and it isn't shutdown for compatible addr,
+        // then won't insert a new one.
+        auto ipv4_any_listener = tcp_listener_map_by_address_.find(
+            Network::Address::Ipv4Instance(address->ip()->port()).asStringView());
+        if (ipv4_any_listener == tcp_listener_map_by_address_.end() ||
+            ipv4_any_listener->second->listener_->listener() == nullptr) {
```

Comment:
Take care of one more case. There could be one "0.0.0.0" listener was stopped, then a "::" with ipv4_compat listener was added, then ensure the ipv6 listener will replace the ipv4 one.